nixos/modules/services/logging/promtail.nix

```
@@ -10,15 +10,29 @@ let
 
   allowPositionsFile = !lib.hasPrefix "/var/cache/promtail" positionsFile;
   positionsFile = cfg.configuration.positions.filename;
+
+  configFile = if cfg.configFile != null
+                   then cfg.configFile
+                   else prettyJSON cfg.configuration;
+
 in {
   options.services.promtail = with types; {
     enable = mkEnableOption "the Promtail ingresser";
 
-
     configuration = mkOption {
       type = (pkgs.formats.json {}).type;
       description = ''
         Specify the configuration for Promtail in Nix.
+        This option is incompatible with `configFile`.
+      '';
+    };
+
+    configFile = mkOption {
+      type = nullOr path;
+      default = null;
+      description = ''
+        Config file path for Promtail.
+        This option is incompatible with `configuration`.
```

Like in the other comment:
```suggestion
        If this option is defined, the value of `services.promtail.configuration` will be ignored.
```